# $\alpha$=0.1

In [2]:
import torch
import numpy as np
import torchvision.transforms as transforms               
from torchvision.datasets import CIFAR10       
from torch.utils.data import DataLoader
import detectors
import timm
from src.temperature_scaling import ModelWithTemperature
from src.aps import aps_test

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# load fine-tuned model
model = timm.create_model("resnet18_cifar10", pretrained=True)
model.to(device)

# reprocess the images from CIFAR10
data_transform = transforms.Compose([
    transforms.ToTensor(),          # transfer to tensor
    transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])  # normalize
])
# load images from CIFAR10
dataset = CIFAR10(root="../../data", train=False, download=True, transform=data_transform)

# Temperature Scaling
temp_scal_loader = DataLoader(dataset, batch_size=32, shuffle=True)
model = ModelWithTemperature(model, temperature=5.0).to(device)
model.set_temperature(temp_scal_loader)

model.eval()

# The number of times the experiment is going to be repeated
num_runs = 10

# error rate
alpha = 0.1

aps_test(model, dataset, device, num_runs=10, alpha=0.1)

Files already downloaded and verified
Before temperature - NLL: 0.293, ECE: 0.039
Optimal temperature: 4.904
After temperature - NLL: 0.866, ECE: 0.464
APS Classification, Start!

Running experiment 1/10...
q_hat = 0.48305028975009917
Total set size: 6225
Total coverage sets: 4476
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.245
Average Coverage Rate in runs 1: 0.8952

Running experiment 2/10...
q_hat = 0.4838371515274048
Total set size: 6338
Total coverage sets: 4500
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.2676
Average Coverage Rate in runs 2: 0.9

Running experiment 3/10...
q_hat = 0.4775569260120392
Total set size: 6109
Total coverage sets: 4446
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.2218
Average Coverage Rate in runs 3: 0.8892

Running experiment 4/10...
q_hat = 0.4783219158649445
Total set size: 6089
Total coverage sets: 4441
Total samples amount: 5000
Average Prediction S

# $\alpha$=0.05

In [3]:
aps_test(model, dataset, device, num_runs=10, alpha=0.05)

APS Classification, Start!

Running experiment 1/10...
q_hat = 0.5326152831315994
Total set size: 7742
Total coverage sets: 4737
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.5484
Average Coverage Rate in runs 1: 0.9474

Running experiment 2/10...
q_hat = 0.5326784014701844
Total set size: 7857
Total coverage sets: 4751
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.5714
Average Coverage Rate in runs 2: 0.9502

Running experiment 3/10...
q_hat = 0.5268210053443909
Total set size: 7581
Total coverage sets: 4721
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.5162
Average Coverage Rate in runs 3: 0.9442

Running experiment 4/10...
q_hat = 0.5255046606063843
Total set size: 7460
Total coverage sets: 4704
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 1.492
Average Coverage Rate in runs 4: 0.9408

Running experiment 5/10...
q_hat = 0.532387575507164
Total set size: 7727

# Result
$\alpha$=0.1
- Final Average **Prediction Set Size： 1.25**
- Final Average **Coverage: 89.62%**  

$\alpha$=0.05
- Final Average **Prediction Set Size： 1.54**
- Final Average **Coverage: 94.69%**